iceoryx2-cal/src/event/common.rs

```diff
@@ -0,0 +1,523 @@
+    {
+        fn timeout(mut self, timeout: Duration) -> Self {
+            self.creation_timeout = timeout;
+            self
+        }
+
+        fn open(
+            self,
+        ) -> Result<
+            <EventImpl<Tracker, WaitMechanism, Storage> as crate::event::Event>::Notifier,
+            crate::event::NotifierCreateError,
+        > {
+            let msg = "Failed to open Notifier";
+
+            match Storage::Builder::new(&self.name)
+                .config(&self.config.convert())
+                .timeout(self.creation_timeout)
+                .open()
+            {
+                Ok(storage) => Ok(Notifier {
+                    storage,
+                    _tracker: PhantomData,
+                    _wait_mechanism: PhantomData,
+                }),
+                Err(DynamicStorageOpenError::DoesNotExist) => {
+                    fail!(from self, with NotifierCreateError::DoesNotExist,
+                        "{} since it does not exist.", msg);
+                }
+                Err(DynamicStorageOpenError::VersionMismatch) => {
+                    fail!(from self, with NotifierCreateError::VersionMismatch,
+                        "{} since the version of the existing construct does not match.", msg);
+                }
+                Err(DynamicStorageOpenError::InitializationNotYetFinalized) => {
+                    fail!(from self, with NotifierCreateError::InitializationNotYetFinalized,
+                        "{} since the initialization is after a timeout of {:?} still not finalized..",
+                        msg, self.creation_timeout);
+                }
+                Err(e) => {
+                    fail!(from self, with NotifierCreateError::InternalFailure,
+                        "{} due to an internal failure ({:?}).", msg, e);
+                }
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct Listener<
+        Tracker: IdTracker,
+        WaitMechanism: SignalMechanism,
+        Storage: DynamicStorage<Management<Tracker, WaitMechanism>>,
+    > {
+        storage: Storage,
+        _tracker: PhantomData<Tracker>,
+        _wait_mechanism: PhantomData<WaitMechanism>,
+    }
+
+    impl<
+            Tracker: IdTracker,
+            WaitMechanism: SignalMechanism,
+            Storage: DynamicStorage<Management<Tracker, WaitMechanism>>,
+        > NamedConcept for Listener<Tracker, WaitMechanism, Storage>
+    {
+        fn name(&self) -> &FileName {
+            self.storage.name()
+        }
+    }
+
+    impl<
+            Tracker: IdTracker,
+            WaitMechanism: SignalMechanism,
+            Storage: DynamicStorage<Management<Tracker, WaitMechanism>>,
+        > crate::event::Listener for Listener<Tracker, WaitMechanism, Storage>
+    {
+        fn try_wait(
+            &self,
+        ) -> Result<Option<crate::event::TriggerId>, crate::event::ListenerWaitError> {
+            while unsafe { self.storage.get().signal_mechanism.try_wait()? } {}
```

Comment:
To collect all notifications until no more are available and then return the id from the tracker. Otherwise, it is possible that someone adds multiple times the same id and `timed_wait` in the signal mechanism does not block and signals falsely that nothing was triggered - just because it did not block.

Reply:
I added a test that verifies exactly this behavior.

Reply:
Already done